### PR TITLE
Permission based limits

### PR DESCRIPTION
**Pull request by ikilltheundead**
These code changes add the ability to use `preciousstones.limit.field_name.#`, and adds a config option to enable the feature, but is disabled by default.

If the config option is enabled, players will need the `preciousstones.limit.field_name` node, and the `preciousstones.limit.field_name.x` node, where x is the maximum number they can place. 

There is another option added to fields called max-per-player, which is my quick and dirty way of limiting the number of times `plugin.getPermissionsManager().has()` is called as I cannot think of a way to check what x is without looping through all possible numbers and checking if that permission node is true. (This is something that needs to be improved upon, but it works for now)

`field_name` in the permission node is the `title:` of the field in the config but lowercase, and with `_` instead of spaces.

If `use-permission-based-limits:` is false in the config, then the `max-per-player: X` will be ignored and needs not to be included in the field settings config.

Sorry about the commits being made by BuildTools. I'm still in the process of learning to use GitHub, and I'm still learning plugin development. Thanks :)

### CODE REVIEW

**Comment by xtomyserrax** [comment]
Hey, first thanks for taking your time and trying to help us to improve this plugin! I will deeply check the code later when I have more time but seems fine to me. Have you tested it?

About max-per-player its well done, I have a staff plugin which has a vanish levels system and I basicly do the same, someone must put an end to the loop and its perfectly done with that permission in each field, then server owners can decide the top per player with perms. I would love if we could only do this check once (each time player joins) and store it, but in my case I only do it with vanish so its easy, here we would need to do it with generic amount of fields

The only thing Im not a fan of is the double permission. What does preciousstones.limit.field_name do? (because then you need to also use preciousstones.limit.field_name.x)
Or did preciousstones.limit.field_name already exist?

Is there already a bypass limit permission? So players with this perm will have unlimited stones to place? (Im not aware of how limit was working before)

**Comment by ikilltheundead** [comment]
Yes, there is already a `preciousstones.bypass.limits`, I think a way to improve would be to remove `preciousstones.limit.field_name` and change that to `preciousstones.bypass.limits.field_name` the idea would be that there would be a way for server owners to give players perms to bypass limits on a per-field basis, whereas bypass.limits is a catch-all. 

Edit: For clarification, that bit of logic is counterintuitive and will need to be changed. With this current code, if a player does not have `preciousstones.limit.field_name` then they will not be limited. This should be changed, so that if they have `preciousstones.bypass.limits.field_name` then there will be no limit on that specific field.

**Comment by xtomyserrax** [comment]
Im not familiar with how limits work so will need to check that. I like the idea to have a general permission, so for example preciousstones.bypass.limits would bypass permission of all fields and maybe preciousstones.bypass.limits.field_name of just one field.

Yeah basicly we need to choose a path, if we have a bypass permission or a limit permission. Idk what @NathanWolf thinks about this

**Comment by NathanWolf** [comment]
I haven't looked over the code yet but reading your description, this *sounds* a lot like the existing permission-based limit system.

However, please correct me if I'm wrong- are you aware that you can do something like this

```
  limits:
  - 6
  - 8
  - 10
  - 12
  - 14
```

which defines 5 different permission groups with different limits, and these can be defined per field?

**Comment by ikilltheundead** [comment]
@NathanWolf 
Say we have 2 fields, Home protection (home_field) and City Protection (city_field). 
Example limit config for both:
home_field:
```
  limits:
  - 1
  - 2
  - 3
  - 4
  - 5
```
city_field:
```
  limits:
  - 0
  - 1
  - 2
  - 3
  - 4
```
if a player has `preciousstones.limit1` then they have 1 home_field and 0 city_field.
if they have `preciousstones.limit3` then they have 3 home_field and 2 city_field.
In most cases, this is fine, but what if I want to give players who (for example) link their mc account to discord an extra home_field? Here's what the limit structure would look like in that case:
home_field:
```
  limits:
  - 1
  - 2
  - 3
  - 4
  - 5
  - 2
  - 3
  - 4
  - 5
  - 6
```
city_field:
```
  limits:
  - 0
  - 1
  - 2
  - 3
  - 4
  - 0
  - 1
  - 2
  - 3
  - 4
```
If using a permissions plugin like LuckPerms, admins could set up a track (PEX Ladders if you're familiar, basically just a method of rank progression) and promote a player 1 time to increase rank or 5 times when a player links the discord.
But bottom line all fields share their limit rank, so if a player has `preciousstones.limit25` then they will be limited to the 25th element in `limit:` for every field. 

My goal was to make it so that admins could have separate limit tracks for each field. With the current method, they would need 10 groups on a track, and it's certainly possible to do, but with my method, they could have a track for home_field and city_field and promote them independently of each other.

Edit TL;DR/summary:
The limit array for each field is tied together by index. This means that admins need to have exponentially more groups if they offer players different ways to increase their field limits

File: src/main/java/net/sacredlabyrinth/Phaed/PreciousStones/managers/LimitManager.java [hunk 1]
```diff
@@ -104,25 +108,25 @@ public int getLimit(Player player, FieldSettings fs) {
 
             return -1;
 
-        } else {
+        } else if (plugin.getSettingsManager().isUsePermissionBasedLimits()) {
 
             //If config flag usePermissionBasedLimits set to true!
             /**
              * Player must have preciousstones.limit.field_title
```
Comment (ikilltheundead) [inline src/main/java/net/sacredlabyrinth/Phaed/PreciousStones/managers/LimitManager.java:115]: This comment is wrong

**Review by ikilltheundead — COMMENTED** [review]
adds `preciousstones.bypass.limit.field_name` to bypass field_name limits.

removes the need for players to have `preciousstones.limit.field_name`

**Comment by ikilltheundead** [comment]
@xtomyserrax 
The rest of the code uses bypass permission instead of limit permission, and I've already made changes to reflect that in the code.

**Comment by NathanWolf** [comment]
Thank you for the explanation, makes sense to me. I will have to leave it up to you and @xtomyserrax to really test and approve, but it is fine with me-  especially since it can be turned off.

**Review by NathanWolf — COMMENTED** [review]
As long as the original functionality is maintained and this can be turned off, it seems fine to me.

**Comment by xtomyserrax** [comment]
Yeah I really liked that it is turned off for default. Can you send me a jar so I can test it please? The initial code I read the first time seemed to not have any issues but I want to be sure before merging

**Comment by ikilltheundead** [comment]
Here ya go
[PreciousStones-ikilltheundead-perm-limit.zip](https://github.com/elBukkit/PreciousStones/files/4861643/PreciousStones-ikilltheundead-perm-limit.zip)

File: src/main/java/net/sacredlabyrinth/Phaed/PreciousStones/managers/SettingsManager.java [hunk 2]
```diff
@@ -569,6 +568,8 @@ public boolean isRedstone(Block block) {
         		block.getType().equals(Material.JUKEBOX) ||
         		block.getType().equals(Material.LEVER) ||
                 block.getType().equals(Material.REPEATER) ||
+                block.getType().equals(Material.DAYLIGHT_DETECTOR) ||
```
Comment (ikilltheundead) [inline src/main/java/net/sacredlabyrinth/Phaed/PreciousStones/managers/SettingsManager.java:571]: Fixes #51

Comment (xtomyserrax) [reply src/main/java/net/sacredlabyrinth/Phaed/PreciousStones/managers/SettingsManager.java:571]: Nice thanks! I may have some time tomorrow to check the permission system you made

**Comment by xtomyserrax** [comment]
So I just tested it and it works perfectly. The code seemed clear when I read it but also after the test no issues were found, I tried to fool it but couldnt find a way!

Thank you a lot for taking your time, some people will love this! If you want to help by fixing another bugs or adding some features people are requesting, feel free :P

**Comment by xtomyserrax** [comment]
Could you please add the new system you made to the wiki @ikilltheundead ? https://github.com/marcelo-mason/PreciousStones/wiki/Permissions

Thank you!!!!

**Pull request merged**